File: flutter/jujuba_svg/test/core/commander/jujuba_commander_test.dart
import 'package:jujuba_svg/core/commander/jujuba_commander.dart';
import 'package:jujuba_svg/core/commander/command.dart';
import 'package:jujuba_svg/model/node_coordinate.dart';
import 'package:test/test.dart';

void main() {
  late JujubaCommander commander;

  setUp(() {
    commander = JujubaCommander();
  });

  test('should emit correct JS for UpdateBackgroundColor', () async {
    final command = UpdateBackgroundColor(
      id: 'node1',
      colorHex: '#FFFFFF',
    );

    expectLater(
      commander.stream,
      emits("updateBackgroundColor('node1','#FFFFFF');"),
    );

    commander.execute(command);
  });

  test('should emit correct JS for UpdateStrokeColor', () async {
    final command = UpdateStrokeColor(
      id: 'node2',
      colorHex: '#000000',
    );

    expectLater(
      commander.stream,
      emits("updateStrokeColor('node2','#000000');"),
    );

    commander.execute(command);
  });

  test('should emit correct JS for UpdateStrokeWidth', () async {
    final command = UpdateStrokeWidth(
      id: 'node3',
      widthInPx: 5,
    );

    expectLater(
      commander.stream,
      emits("updateStrokeWidth('node3',5);"),
    );

    commander.execute(command);
  });

  test('should emit correct JS for RemoveNode', () async {
    final command = RemoveNode(id: 'node4');

    expectLater(
      commander.stream,
      emits("removeNode('node4');"),
    );

    commander.execute(command);
  });

  test('should emit correct JS for UpdateRootBackgroundColor', () async {
    final command = UpdateRootBackgroundColor(colorInHex: '#ABCDEF');

    expectLater(
      commander.stream,
      emits("updateRootBackgroundColor('#ABCDEF');"),
    );

    commander.execute(command);
  });

  test('should emit correct JS for AddRoundedImage', () async {
    final command = AddRoundedImage(
      elementId: 'container',
      imageId: 'img1',
      imageUrl: 'http://example.com/image.png',
      widthInPx: 100,
      heightInPx: 200,
      coordinate: NodeCoordinate(x: 50, y: 75),
    );

    expectLater(
      commander.stream,
      emits("addRoundedImage('container','img1','http://example.com/image.png',"
          "'100','200','50.0','75.0');"),
    );

    commander.execute(command);
  });
}

File: kotlin/jujubasvg/src/commonTest/kotlin/com/github/codandotv/jujubasvg/core/commander/JujubaCommanderTest.kt
package com.github.codandotv.jujubasvg.core.commander

import androidx.compose.ui.graphics.Color
import com.github.codandotv.jujubasvg.model.NodeCoordinate
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.yield
import kotlin.test.Test
import kotlin.test.assertEquals

class JujubaCommanderTest {

    @Test
    fun updateBackgroundColor_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(Command.UpdateBackgroundColor(id = "n1", color = Color.White))
        yield()

        assertEquals("updateBackgroundColor('n1','#ffffffff');", emitted.single())
        job.cancel()
    }

    @Test
    fun updateStrokeColor_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(Command.UpdateStrokeColor(id = "n2", color = Color.Black))
        yield()

        assertEquals("updateStrokeColor('n2','#000000ff');", emitted.single())
        job.cancel()
    }

    @Test
    fun updateStrokeWidth_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(Command.UpdateStrokeWidth(id = "n3", widthInPx = 5))
        yield()

        assertEquals("updateStrokeWidth('n3',5);", emitted.single())
        job.cancel()
    }

    @Test
    fun removeNode_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(Command.RemoveNode(id = "n4"))
        yield()

        assertEquals("removeNode('n4');", emitted.single())
        job.cancel()
    }

    @Test
    fun updateRootBackgroundColor_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(Command.UpdateRootBackgroundColor(color = Color.White))
        yield()

        assertEquals("updateRootBackgroundColor('#ffffffff');", emitted.single())
        job.cancel()
    }

    @Test
    fun customCommand_emitsRawJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        val jsCode = "console.log('hello world');"
        commander.execute(Command.CustomCommand(jsCode = jsCode))
        yield()

        assertEquals(jsCode, emitted.single())
        job.cancel()
    }

    @Test
    fun addRoundedImage_emitsCorrectJS() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(
            Command.AddRoundedImage(
                elementId = "container",
                imageId = "img1",
                imageUrl = "https://example.com/image.png",
                widthInPx = 100,
                heightInPx = 200,
                coordinate = NodeCoordinate(x = 10.5f, y = 20.75f),
            )
        )
        yield()

        assertEquals(
            "addRoundedImage(" +
                    "'container'," +
                    "'img1'," +
                    "'https://example.com/image.png'," +
                    "'100'," +
                    "'200'," +
                    "'10.5'," +
                    "'20.75'" +
                    ");",
            emitted.single(),
        )
        job.cancel()
    }

    @Test
    fun execute_multipleCommands_concatenatesWithNewlines() = runTest {
        val commander = JujubaCommander()
        val emitted = mutableListOf<String>()
        val job = backgroundScope.launch {
            commander.command.collect { emitted.add(it) }
        }
        yield()

        commander.execute(
            Command.RemoveNode(id = "n1"),
            Command.RemoveNode(id = "n2"),
        )
        yield()

        assertEquals(
            "removeNode('n1');\nremoveNode('n2');",
            emitted.single(),
        )
        job.cancel()
    }
}
